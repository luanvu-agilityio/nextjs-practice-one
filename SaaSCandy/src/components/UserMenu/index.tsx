'use client';

import React from 'react';
import Link from 'next/link';
import { useSession, signOut } from 'next-auth/react';
import { ChevronDown, CircleUserRound, LogOut, Settings } from 'lucide-react';

// Shadcn Components
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuLabel,
  DropdownMenuTrigger,
} from '@/components/ui/dropdown-menu';

// Components
import { Typography } from '@/components/common';

// Utils & Constants
import { AUTH_ROUTES } from '@/constants';

interface UserMenuProps {
  className?: string;
}

function UserMenu({ className = '' }: Readonly<UserMenuProps>) {
  const { data: session } = useSession();

  if (!session?.user) {
    return null;
  }

  const {
    user: { name, email, image },
  } = session;

  const handleSignOut = async () => {
    try {
      await signOut({
        callbackUrl: AUTH_ROUTES.HOME,
      });
    } catch (error) {
      console.error('Sign out error:', error);
    }
  };

  const displayName = name || email || 'User';
  const firstName = displayName.split(' ')[0];
  const initials = name
    ? name
        .split(' ')
        .map(n => n[0])
        .join('')
        .toUpperCase()
        .slice(0, 2)
    : email?.charAt(0).toUpperCase() || 'U';

  return (
    <div className={className}>
      <DropdownMenu>
        <DropdownMenuTrigger asChild>
          <button
            className='flex items-center gap-3 px-3 py-2 rounded-lg hover:bg-gray-100 transition-colors focus:outline-none focus:ring-2 focus:ring-orange-background focus:ring-offset-2'
            aria-label='User menu'
          >
            {/* User Avatar */}
            {image ? (
              <img
                src={image}
                alt='Profile'
                className='w-8 h-8 rounded-full border border-form-border-color object-cover'
              />
            ) : (
              <div className='w-8 h-8 rounded-full bg-orange-background flex items-center justify-center text-white text-sm font-medium'>
                {initials}
              </div>
            )}

            {/* User Name */}
            <span className='text-md font-semibold text-orange-background hidden sm:block'>
              Welcome, {firstName}
            </span>

            {/* Dropdown Arrow */}
            <ChevronDown className='w-4 h-4 text-gray-500' />
          </button>
        </DropdownMenuTrigger>

        <DropdownMenuContent
          className='w-64 bg-white shadow-lg border border-gray-200 rounded-lg'
          align='end'
          side='bottom'
          sideOffset={8}
          alignOffset={0}
          avoidCollisions={true}
          collisionPadding={16}
        >
          {/* User Info Header */}
          <DropdownMenuLabel className='p-4'>
            <div className='flex items-center gap-3'>
              {image ? (
                <img
                  src={image}
                  alt='Profile'
                  className='w-10 h-10 rounded-full border border-gray-300 object-cover'
                />
              ) : (
                <div className='w-10 h-10 rounded-full bg-orange-background flex items-center justify-center text-white font-medium'>
                  {initials}
                </div>
              )}
              <div className='min-w-0 flex-1'>
                <Typography className='text-sm font-medium text-gray-900 truncate'>
                  {displayName}
                </Typography>
                <Typography className='text-xs text-gray-500 truncate'>
                  {email}
                </Typography>
              </div>
            </div>
          </DropdownMenuLabel>

          {/* Menu Items */}
          <DropdownMenuItem
            asChild
            className='text-orange-background focus:text-orange-background focus:bg-orange-foreground'
          >
            <Link
              href='/account'
              className='flex items-center gap-3 cursor-pointer'
            >
              <CircleUserRound className='w-5 h-5' />
              <span>Account Details</span>
            </Link>
          </DropdownMenuItem>

          <DropdownMenuItem
            disabled
            className='text-orange-background focus:text-orange-background focus:bg-orange-foreground'
          >
            <div className='flex items-center gap-3'>
              <Settings className='w-5 h-5' />
              <span>Settings (Coming Soon)</span>
            </div>
          </DropdownMenuItem>

          <DropdownMenuItem
            onClick={handleSignOut}
            className='text-orange-background focus:text-orange-background focus:bg-orange-foreground'
          >
            <div className='flex items-center gap-3 cursor-pointer'>
              <LogOut className='w-5 h-5' />
              <span>Sign Out</span>
            </div>
          </DropdownMenuItem>
        </DropdownMenuContent>
      </DropdownMenu>
    </div>
  );
}

export default UserMenu;
