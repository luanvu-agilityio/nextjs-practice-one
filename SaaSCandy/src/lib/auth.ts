'use server';

import { redirect } from 'next/navigation';
import { authApi } from '@/api';
import { ROUTES } from '@/constants';
import type { AuthState } from '@/types';

export async function signInAction(formData: FormData): Promise<AuthState> {
  const email = formData.get('email') as string;
  const password = formData.get('password') as string;

  try {
    // Validate with MockAPI
    const response = await authApi.login({ email, password });

    if (!response.user) {
      return {
        error: 'Invalid email or password',
        success: undefined,
        user: undefined,
        credentials: undefined,
      };
    }

    return {
      success: true,
      user: response.user,
      credentials: { email, password },
      error: undefined,
    };
  } catch (error) {
    console.error('Sign in error:', error);
    return {
      error: error instanceof Error ? error.message : 'Sign in failed',
      success: undefined,
      user: undefined,
      credentials: undefined,
    };
  }
}

export async function signUpAction(formData: FormData): Promise<AuthState> {
  const name = formData.get('name') as string;
  const email = formData.get('email') as string;
  const password = formData.get('password') as string;

  try {
    const [firstName, lastName] = name.split(' ');

    const response = await authApi.register({
      email,
      password,
      firstName: firstName || name,
      lastName: lastName || '',
    });

    if (!response.user) {
      return {
        error: 'Registration failed',
        success: undefined,
        user: undefined,
        credentials: undefined,
      };
    }

    return {
      success: true,
      user: response.user,
      credentials: { email, password },
      error: undefined,
    };
  } catch (error) {
    console.error('Sign up error:', error);
    return {
      error: error instanceof Error ? error.message : 'Sign up failed',
      success: undefined,
      user: undefined,
      credentials: undefined,
    };
  }
}

export async function signOutAction() {
  redirect(ROUTES.HOME);
}
