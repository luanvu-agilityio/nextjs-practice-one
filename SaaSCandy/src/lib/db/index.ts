import { drizzle } from 'drizzle-orm/postgres-js';
import postgres from 'postgres';
import * as schema from './schema';

const connectionString = process.env.POSTGRES_URL!;
export const db = drizzle(postgres(connectionString), { schema });
export * from './schema';
