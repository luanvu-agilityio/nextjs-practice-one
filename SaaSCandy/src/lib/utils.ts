/**
 * Utility function for merging Tailwind CSS class names conditionally.
 *
 * - Uses `clsx` for conditional class logic.
 * - Uses `tailwind-merge` to resolve Tailwind class conflicts.
 *
 * @param inputs - List of class values (strings, arrays, objects).
 * @returns A single merged class name string.

 */
import { clsx, type ClassValue } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}
