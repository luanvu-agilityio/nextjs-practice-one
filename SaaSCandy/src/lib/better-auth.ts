import { betterAuth } from 'better-auth';
import { drizzleAdapter } from 'better-auth/adapters/drizzle';
import { twoFactor } from 'better-auth/plugins';
import { Resend } from 'resend';
import { db } from './db';
import {
  ResetPasswordEmail,
  VerificationEmail,
} from '@/constants/email-template';

const resend = new Resend(process.env.RESEND_API_KEY);

export const auth = betterAuth({
  database: drizzleAdapter(db, {
    provider: 'pg',
  }),

  secret: process.env.BETTER_AUTH_SECRET!,
  baseURL: process.env.BETTER_AUTH_URL || 'http://localhost:3000',

  emailVerification: {
    enabled: true,
    sendOnSignUp: true,
    autoSignInAfterVerification: true,
    requireEmailVerification: true,
    sendVerificationEmail: async ({ user, token }) => {
      const verificationUrl = `${process.env.BETTER_AUTH_URL}/email-verification?token=${token}`;

      await resend.emails.send({
        from: process.env.RESEND_FROM_EMAIL || 'onboarding@resend.dev',
        to: user.email,
        subject: 'Verify your email - SaaSCandy',
        html: VerificationEmail({ verificationUrl }),
      });
    },
  },

  emailAndPassword: {
    enabled: true,
    requireEmailVerification: true,
    sendResetPassword: async ({ user, token }) => {
      const resetUrl = `${process.env.BETTER_AUTH_URL}/reset-password?token=${token}`;

      await resend.emails.send({
        from: process.env.RESEND_FROM_EMAIL || 'onboarding@resend.dev',
        to: user.email,
        subject: 'Reset your password - SaaSCandy',
        html: ResetPasswordEmail({ resetUrl }),
      });
    },
  },

  plugins: [
    twoFactor({
      issuer: 'SaaSCandy',
      otpOptions: {
        period: 30,
        digits: 6,
      },
    }),
  ],

  socialProviders: {
    google: {
      clientId: process.env.GOOGLE_CLIENT_ID || '',
      clientSecret: process.env.GOOGLE_CLIENT_SECRET || '',
      enabled: !!process.env.GOOGLE_CLIENT_ID,
    },
    github: {
      clientId: process.env.GITHUB_ID || '',
      clientSecret: process.env.GITHUB_SECRET || '',
      enabled: !!process.env.GITHUB_ID,
    },
  },

  logger: {
    level: 'debug',
  },
});

export type Session = typeof auth.$Infer.Session;
export type User = (typeof auth.$Infer.Session)['user'];
