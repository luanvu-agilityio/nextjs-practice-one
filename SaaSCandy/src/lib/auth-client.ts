/**
 * Client-side authentication utilities using Better Auth for React.
 *
 * - Provides sign-in, sign-up, sign-out, session hooks, and user update functions.
 * - Supports two-factor authentication (2FA) via the twoFactorClient plugin.
 * - Uses the app's base URL from environment variables.
 *
 * Usage:
 *   import { signIn, useSession } from '@/lib/auth-client';
 */
import { createAuthClient } from 'better-auth/react';
import { twoFactorClient } from 'better-auth/client/plugins';

export const authClient = createAuthClient({
  baseURL: process.env.NEXT_PUBLIC_APP_URL || 'http://localhost:3000',
  plugins: [twoFactorClient()],
});

export const {
  signIn,
  signUp,
  signOut,
  useSession,
  getSession,
  updateUser,
  twoFactor: { enable: enableTwoFactor, disable: disableTwoFactor, verifyOtp },
} = authClient;
