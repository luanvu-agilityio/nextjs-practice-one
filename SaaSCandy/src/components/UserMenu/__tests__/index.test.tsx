import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import UserMenu from '../index';
import { useSession } from '@/lib/auth-client';

jest.mock('@/lib/auth-client');
jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: jest.fn(),
    refresh: jest.fn(),
  }),
}));

describe('UserMenu Component', () => {
  const mockSession = {
    user: {
      name: 'John Doe',
      email: 'john@example.com',
      image: null,
    },
  };

  beforeEach(() => {
    (useSession as jest.Mock).mockReturnValue({ data: mockSession });
  });

  it('matches snapshot', () => {
    const { container } = render(<UserMenu />);
    expect(container).toMatchSnapshot();
  });

  it('opens dropdown menu when clicked', async () => {
    const user = userEvent.setup();
    const { getByRole, getByText } = render(<UserMenu />);

    const trigger = getByRole('button', { name: /user menu/i });
    await user.click(trigger);

    expect(getByText('Account Details')).toBeInTheDocument();
    expect(getByText('Sign Out')).toBeInTheDocument();
  });
});
