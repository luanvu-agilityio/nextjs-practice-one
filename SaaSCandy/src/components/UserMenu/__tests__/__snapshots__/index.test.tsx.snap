// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserMenu Component matches snapshot 1`] = `
<div>
  <div
    class=""
  >
    <button
      aria-expanded="false"
      aria-haspopup="menu"
      aria-label="User menu"
      class="flex items-center gap-3 px-3 py-2 rounded-lg hover:bg-gray-100 transition-colors focus:outline-none focus:ring-2 focus:ring-orange-background focus:ring-offset-2"
      data-state="closed"
      id="radix-«r0»"
      type="button"
    >
      <div
        class="w-8 h-8 rounded-full bg-orange-background flex items-center justify-center text-white text-sm font-medium"
      >
        JD
      </div>
      <span
        class="text-lg font-semibold text-orange-background hidden sm:block"
      >
        Welcome, 
        John
      </span>
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-down w-4 h-4 text-gray-background"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m6 9 6 6 6-6"
        />
      </svg>
    </button>
  </div>
</div>
`;
